bugs in "textattack train”

1. .. training issues:
e.g. 
```
textattack train --model-name-or-path albert-base-v2 --dataset snli --per-device-train-batch-size 8 --epochs 5 --learning-rate 1e-5 

:/server/usera/TextAttack ; textattack train --model-name-or-path albert-base-v2 --dataset snli --per-device-train-batch-size 8 --epochs 5 --learning-rate 1e-5 

Iteration:   0%|                                                                         | 0/68769 [00:00<?, ?it/s]/pytorch/aten/src/THCUNN/ClassNLLCriterion.cu:59: ClassNLLCriterion_updateOutput_no_reduce_kernel: block: [0,0,0], thread: [2,0,0] Assertion `cur_target >= 0 && cur_target < n_classes` failed.
/pytorch/aten/src/THCUNN/ClassNLLCriterion.cu:59: ClassNLLCriterion_updateOutput_no_reduce_kernel: block: [0,0,0], thread: [4,0,0] Assertion `cur_target >= 0 && cur_target < n_classes` failed.
/pytorch/aten/src/THCUNN/ClassNLLCriterion.cu:59: ClassNLLCriterion_updateOutput_no_reduce_kernel: block: [0,0,0], thread: [7,0,0] Assertion `cur_target >= 0 && cur_target < n_classes` failed.
....
/miniconda3/envs/textattackenv-master/lib/python3.7/site-packages/textattack/trainer.py", line 505, in training_step
    sample_weights[is_adv_sample] *= self.training_args.alpha
RuntimeError: CUDA error: device-side assert triggered


```

2. model loading issue, e.g., 
``` 
textattack attack --model t5-en-de --recipe seq2sick --num-examples 10


/server/usera/TextAttack ;  textattack attack --model t5-en-de --recipe seq2sick --num-examples 10 
2021-07-22 17:26:22.554522: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
Downloading: 5.42kB [00:00, 1.63MB/s]                                                                
Downloading: 3.10kB [00:00, 995kB/s]                                                                 
Downloading and preparing dataset ted_multi_translate/plain_text (download: 335.91 MiB, generated: 754.37 MiB, post-processed: Unknown size, total: 1.06 GiB) to /server/usera/.cache/huggingface/datasets/ted_multi_translate/plain_text/1.0.0/60acc1806272a2861722952467c3b4ac274f1a41d798f7c64b7e36cb590dfc48...
Dataset ted_multi_translate downloaded and prepared to /server/usera/.cache/huggingface/datasets/ted_multi_translate/plain_text/1.0.0/60acc1806272a2861722952467c3b4ac274f1a41d798f7c64b7e36cb590dfc48. Subsequent calls will reuse this data.
Traceback (most recent call last):
  File "/server/usera/miniconda3/envs/textattackenv/bin/textattack", line 33, in <module>
    sys.exit(load_entry_point('textattack', 'console_scripts', 'textattack')())
  File "/net/server/usera/TextAttack/textattack/commands/textattack_cli.py", line 42, in main
    func.run(args)
  File "/net/server/usera/TextAttack/textattack/commands/attack_command.py", line 24, in run
    model_wrapper = ModelArgs._create_model_from_args(attack_args)
  File "/net/server/usera/TextAttack/textattack/model_args.py", line 252, in _create_model_from_args
    model, model.tokenizer
  File "/net/server/usera/TextAttack/textattack/models/wrappers/huggingface_model_wrapper.py", line 20, in __init__
    ), f"`model` must be of type `transformers.PreTrainedModel`, but got type {type(model)}."
AssertionError: `model` must be of type `transformers.PreTrainedModel`, but got type <class 'textattack.models.helpers.t5_for_text_to_text.T5ForTextToText'>.
(textattackenv) 
```

She ran this command:
```textattack train --model-name-or-path albert-base-v2 --dataset snli --per-device-train-batch-size 8 --epochs 5 --learning-rate 1e-5```
and got this error:
```RuntimeError: CUDA error: device-side assert triggered```

Instead, we should check the number of labels, and if it's not right through `AssertionError: expected 3 labels, got 4` in this case. And also maybe `AssertionError: label -1 not one of [0, 1, 2]`

--- a/textattack/models/wrappers/huggingface_model_wrapper.py
+++ b/textattack/models/wrappers/huggingface_model_wrapper.py
@@ -7,6 +7,8 @@ import torch
 import transformers
 
 import textattack
+from textattack.models.helpers import T5ForTextToText
+from textattack.models.tokenizers import T5Tokenizer
 
 from .pytorch_model_wrapper import PyTorchModelWrapper
 
@@ -18,11 +20,15 @@ class HuggingFaceModelWrapper(PyTorchModelWrapper):
 
     def __init__(self, model, tokenizer):
         assert isinstance(
-            model, transformers.PreTrainedModel
+            model, (transformers.PreTrainedModel, T5ForTextToText)
         ), f"`model` must be of type `transformers.PreTrainedModel`, but got type {type(model)}."
         assert isinstance(
             tokenizer,
-            (transformers.PreTrainedTokenizer, transformers.PreTrainedTokenizerFast),
+            (
+                transformers.PreTrainedTokenizer,
+                transformers.PreTrainedTokenizerFast,
+                T5Tokenizer,
+            ),
         ), f"`tokenizer` must of type `transformers.PreTrainedTokenizer` or `transformers.PreTrainedTokenizerFast`, but got type {type(tokenizer)}."
 
         self.model = model

--- a/textattack/training_args.py
+++ b/textattack/training_args.py
@@ -539,6 +539,42 @@ class _CommandLineTrainingArgs:
             train_dataset.filter_by_labels_(args.filter_train_by_labels)
         if args.filter_eval_by_labels:
             eval_dataset.filter_by_labels_(args.filter_eval_by_labels)
+        # Testing for Coverage of model return values with dataset.
+        num_labels = args.model_num_labels if args.model_num_labels else 2
+
+        # Only Perform labels checks if output_column is equal to label.
+        if (
+            train_dataset.output_column == "label"
+            and eval_dataset.output_column == "label"
+        ):
+
+            train_dataset_labels = train_dataset._dataset["label"]
+
+            eval_dataset_labels = eval_dataset._dataset["label"]
+
+            train_dataset_labels_set = set(train_dataset_labels)
+
+            assert all(
+                label >= 0
+                for label in train_dataset_labels_set
+                if isinstance(label, int)
+            ), f"Train dataset has negative label/s {[label for label in train_dataset_labels_set if isinstance(label,int) and label < 0 ]} which is/are not supported by pytorch.Use --filter-train-by-labels to keep suitable labels"
+
+            assert num_labels >= len(
+                train_dataset_labels_set
+            ), f"Model constructed has {num_labels} output nodes and train dataset has {len(train_dataset_labels_set)} labels , Model should have output nodes greater than or equal to labels in train dataset.Use --model-num-labels to set model's output nodes."
+
+            eval_dataset_labels_set = set(eval_dataset_labels)
+
+            assert all(
+                label >= 0
+                for label in eval_dataset_labels_set
+                if isinstance(label, int)
+            ), f"Eval dataset has negative label/s {[label for label in eval_dataset_labels_set if isinstance(label,int) and label < 0 ]} which is/are not supported by pytorch.Use --filter-eval-by-labels to keep suitable labels"
+
+            assert num_labels >= len(
+                set(eval_dataset_labels_set)
+            ), f"Model constructed has {num_labels} output nodes and eval dataset has {len(eval_dataset_labels_set)} labels , Model should have output nodes greater than or equal to labels in eval dataset.Use --model-num-labels to set model's output nodes."
 
         return train_dataset, eval_dataset
 

